gnucash/import-export/csv-imp/gnc-csv-price-import-settings.cpp

```diff
@@ -0,0 +1,258 @@
+ * the License, or (at your option) any later version.              *
+ *                                                                  *
+ * This program is distributed in the hope that it will be useful,  *
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of   *
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the    *
+ * GNU General Public License for more details.                     *
+ *                                                                  *
+ * You should have received a copy of the GNU General Public License*
+ * along with this program; if not, contact:                        *
+ *                                                                  *
+ * Free Software Foundation           Voice:  +1-617-542-5942       *
+ * 51 Franklin Street, Fifth Floor    Fax:    +1-617-542-2652       *
+ * Boston, MA  02110-1301,  USA       gnu@gnu.org                   *
+\********************************************************************/
+/** @file gnc-csv-price-import-settings.cpp
+    @brief CSV Import Settings
+    @author Copyright (c) 2014 Robert Fewell
+    @author Copyright (c) 2016 Geert Janssens
+*/
+
+#include "gnc-csv-import-settings.hpp"
+#include "gnc-csv-price-import-settings.hpp"
+#include <sstream>
+
+extern "C"
+{
+#include <config.h>
+
+#include <gtk/gtk.h>
+#include <glib/gi18n.h>
+
+#include "Account.h"
+#include "gnc-state.h"
+#include "gnc-ui-util.h"
+}
+
+const std::string settings_type{"PRICE"};
+
+#define CSV_COL_TYPES    "ColumnTypes"
+
+#define CSV_TO_CURR      "PriceToCurrency"
+#define CSV_FROM_COMM    "PriceFromCommodity"
+
+G_GNUC_UNUSED static QofLogModule log_module = GNC_MOD_IMPORT;
+
+preset_vec_price presets_price;
+
+static std::shared_ptr<CsvPriceImpSettings> create_int_no_preset(void)
+{
+    auto preset = std::make_shared<CsvPriceImpSettings>();
+    preset->m_name = get_no_settings();
+    preset->m_settings_type = settings_type;
+
+    return preset;
+}
+
+static std::shared_ptr<CsvPriceImpSettings> create_int_gnc_exp_preset(void)
+{
+    auto preset = std::make_shared<CsvPriceImpSettings>();
+    preset->m_name = get_gnc_exp();
+    preset->m_skip_start_lines = 1;
+
+    /* FIXME date and currency format should still be aligned with export format!
+     * That's currently hard to do, because the export uses whatever the user
+     * had set as global preference.
+    preset->date_active = 0;
+    preset->currency_active = 0;
+    */
+    preset->m_column_types_price = {
+            GncPricePropType::DATE,
+            GncPricePropType::AMOUNT,
+            GncPricePropType::FROM_COMMODITY,
+            GncPricePropType::TO_CURRENCY
+    };
+    return preset;
+}
+
+/**************************************************
+ * find
+ *
+ * find all settings entries in the state key file
+ * based on settings type.
+ **************************************************/
+const preset_vec_price& get_import_presets_price (void)
```

Comment:
Other than a way to get the group prefix and whether or not there are export format presets, this code is identical across the two specializations we currently have. So this is another candidate to move to a common location.

The most logical place would be a public class method of CsvImportSettings. However as the function is used regardless of whether there are CsvImportSettings instances, it should again be declared static. To get the proper group prefix and the correct no_settings and other type specific internal presets, we do need to know the settings type though. So another template is in order:
```
template<class SType> const preset_vector& get_import_presets (void);
```
(Note the use of preset_vector rather than preset_vector_price, explained in another comment).
```